### PR TITLE
Adds Fastly CDN and Data Caching

### PR DESCRIPTION
**Pull request by redreceipt**
### What does this PR do, or why is it needed?

This will add a cache layer to all our queries and a corresponding Fastly CDN route to handle them.

### What design trade-offs/decisions were made?

None.

### How do I test this PR?

- [ ] Make sure most queries are being cached, you can use the Network inspector from `react-native-debugger`

Run queries using the following types and make sure they aren't cached:

- [ ] `AuthenticatedUser`
- [ ] `Person`
- [ ] `userFeed`
- [ ] `isLiked`
- [ ] `likedCount`
- [ ] `getAllLikedContent`

### What automated tests did you write?

None.

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [#494]
- [ ] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [ ] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

File: packages/apollos-data-schema/src/index.js [hunk 1]
```diff
@@ -330,6 +330,7 @@ export const contentItemSchema = gql`
 
   extend type Query {
     userFeed(first: Int, after: String): ContentItemsConnection
+      @cacheControl(maxAge: 0)
```
Comment (conrad-vanl) [inline packages/apollos-data-schema/src/index.js]: Is there a way to cache this by user?

**Comment by vinnyjth** [comment]
> Does CDN caching handle partial query caching or just full queries? If just full, how would we implement partial query caching?

Full query caching only :/ @redreceipt and I talked about the need to comb the client for queries that use fields like "isLiked" and make sure that the fields that eliminate the ability to cache aren't included in cacheable queries 

For partial query caching, we need to either 
- add Max-Age headers to the data coming from Rock (so that the datasources will cache) 
- or add `.ttl` calls to all of our rock requests (and make sure that the `ttl` option is working  
- or Roll our own caching layer that is capable of handling partial query caching. There's no out of the box options at this time to do partial query caching with a cdn.

File: packages/apollos-data-schema/src/index.js (hunk 1 above)
Comment (vinnyjth) [reply packages/apollos-data-schema/src/index.js]: @conrad-vanl Yes! Micheal and I talked about this yesterday. We need to change `scope: PRIVATE` and then write a cache policy in Fastly. I'm working with @redreceipt to figure out how to accomplish that in this PR (probably need to add some documentation about the Fastly side of things)

**Comment by redreceipt** [comment]
@vinnyjth ok trying to test out what we've got, definitely getting some network errors.
screen shot 2019-02-04 at 9 37 07 am
screen shot 2019-02-04 at 9 37 12 am

File: packages/apollos-data-schema/src/index.js [hunk 2]
```diff
@@ -415,12 +416,12 @@ export const followingsSchema = gql`
   }
 
   ${extendForEachContentItemType(`
-    isLiked: Boolean
-    likedCount: Int
+    isLiked: Boolean @cacheControl(maxAge: 0)
+    likedCount: Int @cacheControl(maxAge: 0)
```
Comment (vinnyjth) [inline packages/apollos-data-schema/src/index.js]: I don't think we need to cache `likedCount` because it should be the same for all users (and it's okay if it's a little slow to update)

Comment (conrad-vanl) [reply packages/apollos-data-schema/src/index.js]: Are we able to manually invalidate the cache for `likedCount` whenever a `liked` mutation is fired for that content item? I could see a problem where a user taps "like" and the like counter in app doesn't update...especially if it still shows "0 likes"

Comment (vinnyjth) [reply packages/apollos-data-schema/src/index.js]: Not without pulling in the Fastly API and manually flushing the cache when an item is liked :/

However, we could update the code here 
https://github.com/ApollosProject/apollos-prototype/blob/master/packages/apolloschurchapp/src/ui/LikeButton/index.js
to increase the `likedCount` for the item on the client. It would fall back to an old value if the view was refetched, but hopefully by that time the cache has expired. 

Another option, we could dynamically add a `expires-at` header based on the number of likes using this API. We could set a max-age of 0 for content items with no likes, and start adding a cache at a likedCount of 10 or so.
```
const resolvers = {
  Query: {
    post: (_, { id }, _, { cacheControl }) => {
      cacheControl.setCacheHint({ maxAge: 60 });
      return find(posts, { id });
    }
  }
}
```

**Comment by redreceipt** [comment]
@vinnyjth

I think figuring out `scope` takes care of 1 and 3. I'll make an issue for this.

I do have a documentation skeleton up and running, I'm going to make some finishing touches to the site, including instructions for Fastly, and get that up so we can all make edits as we go.

For 4, I could be thinking in the wrong direction completely but Fastly has an API that we could maybe use to flush certain queries?

**Comment by vinnyjth** [comment]
Sounds good :D Once you create those issues reference this issues to those issues and I'll close out this PR.

Do you have a link to what you have right now in terms of documentation? 

I think that's a good direction to explore. A frustration with that approach is that it would increase our dependency on Fastly. Another problem will be that actually identifying a record in the cache will be very difficult. For example, the cache key for the isLive query is a hashed version of this. 

http://localhost:4000/?query=query%20getLiveStream%20%7B%0A%20%20liveStream%20%7B%0A%20%20%20%20isLive%0A%20%20%20%20__typename%0A%20%20%7D%0A%7D%0A&operationName=getLiveStream&variables=%7B%7D

if we ever change that query, the cache key will change. If any other queries include that sensitive field, we will need to somehow identify those queries in the cache. There might be a be a good way to do this, but we will be heavily dependent of the power of the Fastly tools for searching through data that's cached.

**Comment by vinnyjth** [comment]
@redreceipt One more thing before we merge, can you change the value in `.env.production` to the fastly URL? Make sure we include `https://` in the url.

**Comment by redreceipt** [comment]
@vinnyjth 

Issue #505 is linked.

The documentation is in the `ApollosProject.github.io` repo if you wanna poke around. We'll create new branches off of the `source` branch for making changes. We'll publish from the `source` branch which will push to `master` and go live. Haven't done that step yet, almost ready.

`.env.production` has been updated to Fastly URL.

File: packages/apollos-data-schema/src/index.js (hunk 2 above)
Comment (vinnyjth) [reply packages/apollos-data-schema/src/index.js]: @conrad-vanl after thinking about this some more, we opted to keep the cache exception for this field for the time being.
